src/Adapter/Product/Repository/ProductRepository.php

```diff
@@ -357,4 +362,49 @@ public function bulkDelete(array $productIds): void
             sprintf('Failed to delete following products: "%s"', implode(', ', $failedIds))
         );
     }
+
+    /**
+     * @param ProductId $productId
+     * @param ShopId $shopId
+     *
+     * @return bool
+     */
+    public function isAssociatedToShop(ProductId $productId, ShopId $shopId): bool
+    {
+        try {
+            $this->checkShopAssociation($productId->getValue(), Product::class, $shopId);
+        } catch (ShopAssociationNotFound $e) {
+            return false;
+        }
```

Comment:
You shouldn't use exceptions as a messaging protocol

Reply:
I suppose @eternoendless says that you use the throw of ShopAssociationNotFound to fetch the data "there is no shop association" and return false. I suppose @eternoendless would prefer

```
if (!$this->isThereAShopAssociation(...) {
    return false;
}

$check = $this->checkShopAssociation(...);

return true;
```

Don't rely on "this code will throw exception B if it meets usecase A, so I will on purpose trigger the situation for usecase A and listen for exception B to know whether or not I am in usecase A". Exceptions should only be throwed when we go on a path we DONT want to go.